lib/std/mem.zig

```diff
@@ -876,6 +874,63 @@ pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, nee
     return null;
 }
 
+fn boyerMooreHorspoolPreprocess(pattern: []const u8, table: []usize) void {
+    for (table) |*c| {
+        c.* = pattern.len;
+    }
+
+    var i: usize = 0;
+    while (i < pattern.len - 1) : (i += 1) {
+        table[pattern[i]] = pattern.len - 1 - i;
+    }
+}
+/// Find the index in a slice of a sub-slice, searching from the end backwards.
+/// To start looking at a different index, slice the haystack first.
+// Reverse boyer-moore-horspool algorithm
+pub fn lastIndexOf(comptime T: type, haystack: []const T, needle: []const T) ?usize {
+    if (haystack.len < 32 or needle.len <= 2)
+        return lastIndexOfNaive(T, haystack, needle);
+
+    if (needle.len > haystack.len or needle.len == 0) return null;
+
+    const haystackU8 = @bitCast([]const u8, haystack);
+    const needleU8 = @bitCast([]const u8, needle);
+
+    var table: [256]usize = undefined;
+    boyerMooreHorspoolPreprocess(needleU8, table[0..]);
+
+    var i: usize = 0;
+    while (i <= haystackU8.len - needleU8.len) {
+        const reverseIndex = haystackU8.len - i - needleU8.len - 1;
+        if (mem.eql(u8, haystackU8[reverseIndex .. reverseIndex + needleU8.len], needleU8)) return i;
+        i += table[haystackU8[reverseIndex + needleU8.len - 1]];
+    }
+
+    return null;
+}
+
+// Boyer-moore-horspool algorithm
+pub fn indexOfPos(comptime T: type, haystack: []const T, start_index: usize, needle: []const T) ?usize {
+    if (haystack.len < 32 or needle.len <= 2)
+        return indexOfPosNaive(T, haystack, start_index, needle);
+
+    if (needle.len > haystack.len or needle.len == 0) return null;
+
+    const haystackU8 = @bitCast([]const u8, haystack);
```

Comment:
@andrewrk wrote this here https://github.com/ziglang/zig/issues/1060#issuecomment-614263132:

> > Is it valid to cast these directly to a byte representation and search on that?
>
> no - this is valid only for types which have well-defined memory layout

I agree - this should only work for things like packed structs, slices of integers etc.

Reply:
I think this was also mentioned in https://github.com/ziglang/zig/pull/6145: Floating point numbers `+0.0` and `-0.0` are considered equal, but their byte representation is not equal.

Reply:
Hmm i added a check that the type is a power of 2 not but not float. What would be the best way to make sure it works for all cases?

Reply:
Ok so I added a check now if its float or not. What other types of checks are needed? if its a pod?